File: common/src/main/java/generations/gg/generations/core/generationscore/common/world/recipe/GenerationsCoreRecipeTypes.java
package generations.gg.generations.core.generationscore.common.world.recipe;

import dev.architectury.registry.registries.DeferredRegister;
import dev.architectury.registry.registries.RegistrySupplier;
import generations.gg.generations.core.generationscore.common.GenerationsCore;
import net.minecraft.core.registries.Registries;
import net.minecraft.world.item.crafting.Recipe;
import net.minecraft.world.item.crafting.RecipeType;

public class GenerationsCoreRecipeTypes {
	public static final DeferredRegister<RecipeType<?>> RECIPES_TYPES = DeferredRegister.create(GenerationsCore.MOD_ID, Registries.RECIPE_TYPE);
	public static RegistrySupplier<RecipeType<RksRecipe>> RKS = register("rks");

	private static <T extends Recipe<?>> RegistrySupplier<RecipeType<T>> register(String name) {
		var id = GenerationsCore.id(name);
		return RECIPES_TYPES.register(id, () -> new RecipeType<>() {
            @Override
            public String toString() {
                return id.toString();
            }
        });
	}

    public static void init() {
		RECIPES_TYPES.register();
	}
}

File: common/src/main/java/generations/gg/generations/core/generationscore/common/world/recipe/GenerationsCoreRecipeTypes.kt
package generations.gg.generations.core.generationscore.common.world.recipe

import generations.gg.generations.core.generationscore.common.generationsResource
import generations.gg.generations.core.generationscore.common.util.PlatformRegistry
import net.minecraft.core.Holder
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.core.registries.Registries
import net.minecraft.world.item.crafting.Recipe
import net.minecraft.world.item.crafting.RecipeType

object GenerationsCoreRecipeTypes: PlatformRegistry<RecipeType<*>>(Registries.RECIPE_TYPE, BuiltInRegistries.RECIPE_TYPE) {

    @JvmField
	var RKS = register<RksRecipe>("rks")

    private fun <T : Recipe<*>> register(name: String): Holder<RecipeType<*>> {
        val id = name.generationsResource()
        return create(name, {
            object : RecipeType<T> {
                override fun toString(): String = id.toString()
            }
        })
    }
}
